### PR TITLE
More mapping table improvements

### PR DESCRIPTION
**Pull request by aeschli**
Here are more performance improvements around utf-8 mapping:

- based on how large the urf-8 indexes can be (maximum is utf-16.length * 3), create a UInt8/ 16/ 32 Array 
- create the mapping table starting from the offset of the first multi byte character, and normalize the urf-8 indexes by that index
   - `_mappingTableStartOffset`: utf16-offset where the mapping table starts. Before that index: utf16-index === utf8-index
   - `_utf8Indexes`: utf-16 to utf-8 mapping table for all uft-8 indexes starting at `_mappingTableStartOffset`. utf8-index are always starting at 0.
   - Example: _mappingTableStartOffset === 10, _utf8Indexes = [0, 3, 6] 
-> utf8 index at 10 = 10, utf8 index at 11 = 13

### CODE REVIEW

**Comment by aeschli** [comment]
```
Uncached strings < 100 characters
: 112 ms
Cached strings < 100 characters
: 17 ms

Uncached strings < 1000 characters
: 333 ms
Cached strings < 1000 characters
: 31 ms

Uncached strings < 10000 characters
: 2705 ms
Cached strings < 10000 characters
: 169 ms

Uncached strings < 50000 characters
: 16949 ms
Cached strings < 50000 characters
: 800 ms

Uncached strings < 100000 characters
: 35859 ms
Cached strings < 100000 characters
: 1599 ms
```

**Comment by mofux** [comment]
@aeschli Is this an effort to enhance the default monaco-editor highlighter with textmate grammars? I'm about to write an implementation by myself, but I am wondering if I should hold back if official support is being considered.

**Comment by NeekSandhu** [comment]
@aeschli  Please accept collaborator invite for write access

**Comment by aeschli** [comment]
@mofux Yes that's the goal

**Comment by aeschli** [comment]
Thanks for the write access, @NeekSandhu. Is ok for you if I merge this PR?

**Comment by NeekSandhu** [comment]
After #10 I'm totally blown away by the fact that how can dynamic lookup (binary search) be faster (overall) than generating an offset table once and looking up offset **directly** 😲

Haven't slept for days thinking about it.

That being said, #10 taught me new ways of looking at things in future. Thanks to you for that.

Please feel free to merge this one and let's consider conditional handling based on `str.length` as this PR improves some and slows down other cases in `benchmark.js` as compared to #10. Once again hard to wrap head around this behaviour.

File: src/OnigString.ts [hunk 1]
```diff
@@ -190,10 +230,18 @@ function findFirstInSorted<T>(array: Int32Array, i: number): number {
             low = mid + 1
         }
     }
-    if (array[low] > i) {
-        low--
+
+    // low is on the index of the first value >= i or array.length. Decrement low until we find array[low] <= i 
+    while (low > 0 && (low >= array.length || array[low] > i)) {
```
Comment (NeekSandhu) [inline src/OnigString.ts]: I have a feeling that binary-tree like search might be even faster. Did you try that already?

Comment (aeschli) [reply src/OnigString.ts]: Not sure what you mean by binary search tree like search.
What we do here is doing is a [Binary_search](https://en.wikipedia.org/wiki/Binary_search_algorithm) for the first entry in the table with value >= x. After that we just step back if we only found a value > x (that's normally just one step back, or two, if we're on a UTF-16 surrogate char. (the 'while' is misleading here)

Comment (NeekSandhu) [reply src/OnigString.ts]: I imagine we're searching by decrementing one each time (`low--`) as opposed to halving the search block; rinse and repeat until found.

>  Binary search compares the target value to the middle element of the array; if they are unequal, the half in which the target cannot lie is eliminated and the search continues on the remaining half until the target value is found

Comment (NeekSandhu) [reply src/OnigString.ts]: Please see [text-store/index.ts#L132-L150](https://github.com/chipto/text-store/blob/70b347392b9207841a0bcabda0112664db4dff58/src/index.ts#L132-L150)

Comment (NeekSandhu) [reply src/OnigString.ts]: Maybe I can give it a shot sometime later and open a new PR if it does any more good

Comment (aeschli) [reply src/OnigString.ts]: Not sure if I fully understood what you meant, but yes, using the previous search result as a starting point for the next search is a great idea.
But I suggest we look at that in a next PR.

**Comment by aeschli** [comment]
Merged. Thanks for the review.

**Comment by NeekSandhu** [comment]
Thank you so much for your contribution Martin!

Patch is now available in [2.2.0](https://github.com/NeekSandhu/onigasm/releases/tag/v2.2.0)

**Pull request merged**